lib/bigdecimal.rb

```diff
@@ -3,3 +3,263 @@
+
+    if y < 0
+      inv = x.power(-y, prec)
+      return BigDecimal(0) if inv.infinite?
+      return BigDecimal::INFINITY if inv.zero?
+      return BigDecimal(1).div(inv, prec)
+    end
+
+    int_part = y.fix.to_i
+    prec2 = prec + BigDecimal.double_fig
+    pow_prec = prec2 + (int_part > 0 ? y.exponent : 0)
+    ans = BigDecimal(1)
+    n = 1
+    xn = x
+    while true
+      ans = ans.mult(xn, pow_prec) if int_part.allbits?(n)
+      n <<= 1
+      break if n > int_part
+      xn = xn.mult(xn, pow_prec)
+    end
+    unless frac_part.zero?
+      ans = ans.mult(BigMath.exp(BigMath.log(x, prec2).mult(frac_part, prec2), prec2), prec2)
+    end
+    ans.mult(1, prec)
+  end
+end
+
+# Core BigMath methods for BigDecimal (log, exp) are defined here.
+# Other methods (sin, cos, atan) are defined in 'bigdecimal/math.rb'.
+module BigMath
+  def self._coerce_to_bigdecimal(x, method_name, complex_domain_error = false)
+    case x
+    when BigDecimal
+      return x
+    when Integer, Float, Rational
+      return BigDecimal(x, 0)
+    when Complex
+      if complex_domain_error
+        raise Math::DomainError, "Complex argument for BigMath.#{method_name}"
+      end
+    end
+    raise ArgumentError, "#{x.inspect} can't be coerced into BigDecimal"
+  end
+
+  def self._validate_prec(prec, method_name)
+    raise ArgumentError, 'precision must be an Integer' unless Integer === prec
+    raise ArgumentError, "Zero or negative precision for #{method_name}" if prec <= 0
+  end
+
+  # call-seq:
+  #   BigMath.log(decimal, numeric)    -> BigDecimal
+  #
+  # Computes the natural logarithm of +decimal+ to the specified number of
+  # digits of precision, +numeric+.
+  #
+  # If +decimal+ is zero or negative, raises Math::DomainError.
+  #
+  # If +decimal+ is positive infinity, returns Infinity.
+  #
+  # If +decimal+ is NaN, returns NaN.
+  #
+  def self.log(x, prec)
+    _validate_prec(prec, :log)
+    x = _coerce_to_bigdecimal(x, :log, true)
+    return BigDecimal::NAN if x.nan?
+    raise Math::DomainError, 'Zero or negative argument for log' if x <= 0
+    return BigDecimal::INFINITY if x.infinite?
+    return BigDecimal(0) if x == 1
+
+    if x > 10 || x < 0.1
+      log10 = log(BigDecimal(10), prec)
+      exponent = x.exponent
+      x = x * BigDecimal("1e#{-x.exponent}")
+      if x > 3
+        x /= 10
+        exponent += 1
+      end
+      return log10 * exponent + log(x, prec)
+    end
+
+    x_minus_one_exponent = (x - 1).exponent
+    prec += BigDecimal.double_fig
+
+    # log(x) = log(sqrt(sqrt(sqrt(sqrt(x))))) * 2**sqrt_steps
+    sqrt_steps = [2 * Integer.sqrt(prec) + 3 * x_minus_one_exponent, 0].max
+
+    # Reduce sqrt_step until sqrt gets fast
+    # https://github.com/ruby/bigdecimal/pull/323
+    # https://github.com/ruby/bigdecimal/pull/343
+    sqrt_steps /= 10
+
+    prec2 = prec + [-x_minus_one_exponent, 0].max + (sqrt_steps * 0.3010299956639812).ceil
```

Comment:
`0.3010299956639812 == Math.log10(2)`
`x = sqrt(x)` makes `x-1` about half. (`sqrt(1+delta) ≒ 1+delta/2`)
performing sqrt 10 times makes `x-1` about 1/1024.
`sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(sqrt(1.0001)))))))))) ≒ 1.0000001`
The calculation result requires `log10(2**10)` extra digits.